### DISCUSSION TITLE
How to replace DefaultKeycloakSession usage with custom realization?

### DISCUSSION THREAD

**Discussion by MelTrevelyan** [keycloak/keycloak#34225] [Q&A]
Hi guys! 
I work with 24.0.5 version. The thing is, I need to add additional logging to the `KeycloakSession.close()` method. The DefaultKeycloakSession provides logs on trace level, but dont need trace in terms of other parts of the service. I tried to create an SPI with CustomKeycloakSession that works like a wrapper to KeycloakSession delegate and adds additional logs, but I don`t know how to tell keycloak to use my version.

Can you help me please? Or maybe there is another solution to add logs in this case? Thanks!

**Comment by danielFesenmeyer** [comment]
It seems like you cannot simply inject another KeycloakSession implementation.
Keycloak uses the static method QuarkusKeycloakSessionFactory#getInstance at several places. (Keycloak no longer uses DefaultKeycloakSession, it uses QuarkusKeycloakSession, at least in production).

Maybe someone else has an idea how such logging can be added?

**Comment by shawkins** [comment]
Do you need logging for every session, or just the main session used for the request processing lifecycle? If it's the latter, then you can add another filter, like the builtinnn close filter, but that runs before it https://github.com/keycloak/keycloak/blob/358d234538d242568628fe7dad5811de34e2ffca/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/jaxrs/CloseSessionFilter.java#L36

That is not a supported mechanism though.
